Packaging and Exporting with Cmake

AlexBuccheri opened the pull request:
Implementation of project installation, such that one can use it with `find_package` and `fetchContent`

AlexBuccheri committed:
Make project installable and exportable

AlexBuccheri committed:
Template for pkgconfig

AlexBuccheri commented on CMakeLists.txt:
```diff
@@ -2,7 +2,14 @@ cmake_minimum_required(VERSION 3.20)
-set(CMAKE_Fortran_MODULE_DIRECTORY ${PROJECT_BINARY_DIR}/modules)
+# Note, this could also be stored in version text: configure_file(version.txt.in version.txt)
+# or dynamically extracted from git tags
+set(libkmeans_VERSION_MAJOR 1)
+set(libkmeans_VERSION_MINOR 0)
+set(libkmeans_VERSION_PATCH 0)
```
Better to set PROJECT_VERSION and define from that?

LecrisUT replied:
Put it directly in `project(VERSION)`. If you ever see stuff like that it is because of autotools "conversion".

Re: dynamaically extracted: https://github.com/LecrisUT/CMakeExtraUtils/blob/main/cmake/DynamicVersion.md

AlexBuccheri replied:
>  https://github.com/LecrisUT/CMakeExtraUtils/blob/main/cmake/DynamicVersion.md


Super nice, but I won't be using tags in this project. It's just not big enough in scope.

AlexBuccheri commented on CMakeLists.txt:
```diff
@@ -146,3 +154,58 @@ target_link_libraries(test_kmeans PRIVATE lib_kmeans fortuno_wrapper common_flag
+   DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/kmeans"
+   )
+
+# Module files.
+install(
+   DIRECTORY ${CMAKE_Fortran_MODULE_DIRECTORY}
+   DESTINATION "."  # Will automatically get placed in "${CMAKE_INSTALL_LIBDIR}/modules"
+ )
```
I note that no modules currently install, perhaps since I changed lib_kmeans src to PRIVATE?

LecrisUT replied:
Partly because you have removed:
```cmake
set(CMAKE_Fortran_MODULE_DIRECTORY ${PROJECT_BINARY_DIR}/modules)
```

The comment there is wrong. It would be installing to `/usr` a.k.a. `CMAKE_INSTALL_PREFIX`. Do specify the path. But note the `/` which would either install everything in the directory or install the directory itself.

The install path is also annoying because it has to be project specific (and compiler specific sometimes).

AlexBuccheri replied:
Weird, didn't get comment notifications come through on this.

> Partly because you have removed ..

Yeah, I just noticed that.

>  It would be installing to /usr a.k.a. CMAKE_INSTALL_PREFIX

Sorry, that was implicit in my comment (to me at least)

Ahhh, it's copying the directory, not the contents! No I get it. Thanks

AlexBuccheri commented on cmake/kmeans.pc.in:
```diff
@@ -0,0 +1,8 @@
+includedir=@CMAKE_INSTALL_FULL_LIBDIR@/modules 
+
+Cflags: -I@CMAKE_INSTALL_FULL_LIBDIR@/modules
+Libs: -L@CMAKE_INSTALL_FULL_LIBDIR@ -lkmeans @MPI_LIBRARIES@ -lmpi_usempif08
```
No idea if I want to include `lmpi_usempif08`. I assume yes, but that it should be conditional on the build options. This file resolves to:

```
Name: libkmeans
Description: 
Version: 1
libdir=/Users/alexanderbuccheri/Codes/kmeans/install/lib
includedir=/Users/alexanderbuccheri/Codes/kmeans/install/lib/modules 

Cflags: -I/Users/alexanderbuccheri/Codes/kmeans/install/lib/modules
Libs: -L/Users/alexanderbuccheri/Codes/kmeans/install/lib -lkmeans  -lmpi_usempif08
```

except there currently isn't be a modules folder

LecrisUT replied:
This is the whole point why I hate `.pc` files. What you can try is to chain `configure_file` (replaces `@@`) -> `file(GENERATE)` (replaces genex `$<>`)

You can try:
```
Name: @PROJECT_NAME@
Description: @PROJECT_DESCRIPTION@
Version: @PROJECT_VERSION@
prefix=@CMAKE_INSTALL_PREFIX@
libdir=${prefix}/@CMAKE_INSTALL_LIBDIR@
includedir=${prefix}/@CMAKE_INSTALL_INCLUDEDIR@
moddir=${prefix}/@CMAKE_INSTALL_LIBDIR@/modules

Cflags: -I${includedir}
Libs: -L${libdir} -l$<TARGET_LINKER_LIBRARY_FILE_BASE_NAME:lib_kmeans> $<TARGET_PROPERTY:lib_kmeans,INTERFACE_LINK_LIBRARIES>
```

What must happen is that all `target_link_libraries(PRIVATE)` are not included here, and all `target_link_libraries(PUBLIC/INTERFACE)` are included. This list depends on what feature you enable at the build time and how `find_package` determined the flags necessary.

My honest recommendation, just don't bother with `.pc`

LecrisUT replied:
[Here](https://sourcegraph.com/github.com/scylladb/seastar@998d29970b9207c4ba8a125684f883c363d1010a/-/blob/pkgconfig/seastar.pc.in) is an example of how messed up it can be to support that. If you want more inspiration:
https://sourcegraph.com/search?q=context:global+file:.*%5C.pc.*+%24%3CTARGET_PROPERTY&patternType=keyword&sm=0

AlexBuccheri commented on CMakeLists.txt:
```diff
@@ -146,3 +154,58 @@ target_link_libraries(test_kmeans PRIVATE lib_kmeans fortuno_wrapper common_flag
+    FILES
+        "${CMAKE_CURRENT_BINARY_DIR}/kmeansConfigVersion.cmake"
+    DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/kmeans"
+    COMPONENT dev
```
I didn't really understand what this does

LecrisUT replied:
`COMPONENT`? It's used on [`cmake --install --component <comp>`](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-install-component). Follow good practices before the angry CMake mob comes knocking on the door.

AlexBuccheri replied:
Ok, I've removed entirely as I was not exposing any components.

LecrisUT commented on CMakeLists.txt:
```diff
@@ -2,7 +2,14 @@ cmake_minimum_required(VERSION 3.20)
+set(libkmeans_VERSION_MINOR 0)
+set(libkmeans_VERSION_PATCH 0)
+
+list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake)
```
```suggestion
```
Try not to set it if you don't use it.

LecrisUT commented on CMakeLists.txt:
```diff
@@ -79,21 +86,23 @@ endif()
+    LIBRARY_OUTPUT_NAME "kmeans"
+    ARCHIVE_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}"
+    LIBRARY_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}"
+    Fortran_MODULE_DIRECTORY "${CMAKE_Fortran_MODULE_DIRECTORY}"
```
```suggestion
```

Redundant. `CMAKE_Fortran_MODULE_DIRECTORY` -> initialize the property `Fortran_MODULE_DIRECTORY`

LecrisUT commented on CMakeLists.txt:
```diff
@@ -146,3 +154,58 @@ target_link_libraries(test_kmeans PRIVATE lib_kmeans fortuno_wrapper common_flag
+install(
+   TARGETS lib_kmeans
+   EXPORT KmeansTargets
+   LIBRARY
+     DESTINATION "${CMAKE_INSTALL_LIBDIR}"
+   ARCHIVE
+     DESTINATION "${CMAKE_INSTALL_LIBDIR}"
```
Best if you keep them empty. It should default to the GnuInstallDir iiuc.

LecrisUT commented on CMakeLists.txt:
```diff
@@ -102,13 +111,12 @@ target_link_libraries(lib_kmeans PRIVATE
    common_flags 
    )
```
```suggestion
   )
target_include_directories(lib_kmeans PUBLIC
        "$<BUILD_INTERFACE:${CMAKE_Fortran_MODULE_DIRECTORY}>"
        "$<INSTALL_INTERFACE:${CMAKE_INSTALL_MODULEDIR}>"
)
```

You have to specify different `include` directives on build and install

AlexBuccheri replied:
Why do I need this? Aren't I specifying these directives elsewhere?

LecrisUT replied:
You have to specify how the CMake consumer should consume `lib_kmeans` (even within the same project). And no, you do not have it specified anywhere. It just accidentally works because everything is built in the same `CMAKE_Fortran_MODULE_DIRECTORY` which is implicitly included. This does not work when other projects try to use your target.

AlexBuccheri replied:
hmm, this is new to my limited cmake installation knowledge. I've now added it, but it's not intuitive (for me)

LecrisUT commented on CMakeLists.txt:
```diff
@@ -146,3 +154,58 @@ target_link_libraries(test_kmeans PRIVATE lib_kmeans fortuno_wrapper common_flag
+# export(TARGETS lib_kmeans
+#    FILE "${CMAKE_CURRENT_BINARY_DIR}/kmeansTargets.cmake"
+#    NAMESPACE kmeans::
+# )
```
This is unintuitive, but `export(TARGETS)` creates the `kmeansTargets.cmake` used for the build folder, i.e. you can use `find_package(kmeans)` and point `kmeans_ROOT=${binary-dir}` and you can test 90% of the packaging there.

It should *always* be included

AlexBuccheri committed:
Note issues with .pc template, and disable.
Fix how versioning variables are set.
Fix installation of module files.
Clean up CMakeLists.txt structure.

LecrisUT commented on CMakeLists.txt:
```diff
@@ -74,26 +91,26 @@ else()
-        LIBRARY_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}
-        )
+set_target_properties(lib_kmeans PROPERTIES
+    VERSION "${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}.${PROJECT_VERSION_PATCH}"
```
```suggestion
    VERSION "${PROJECT_VERSION}"
```
Why write is simple am I right?

AlexBuccheri replied:
😅 I wasn't focused on this - yes, clearly simplification now `VERSION` is defined for the project

LecrisUT commented on CMakeLists.txt:
```diff
@@ -146,3 +167,66 @@ target_link_libraries(test_kmeans PRIVATE lib_kmeans fortuno_wrapper common_flag
+# Note, this must be consistent with cmake/kmeans.pc.in
+install(
+   DIRECTORY ${CMAKE_Fortran_MODULE_DIRECTORY}
+   DESTINATION "${CMAKE_INSTALL_PREFIX}"    #Copies directory ${CMAKE_Fortran_MODULE_DIRECTORY}, to this location
```
Install to `/usr/modules`? I guess you-do-you

AlexBuccheri replied:
`${CMAKE_INSTALL_PREFIX}` I thought that's the default, which gets overwritten if I configure with

```shell
cmake --fresh -B serial-cmake-build-debug -DCMAKE_BUILD_TYPE=Debug -DMPI=Off -DCMAKE_INSTALL_PREFIX=/Users/alexanderbuccheri/Codes/kmeans/install
```

Seems to be what's happening:

```shell
(.venv) (base) Administrators-MacBook-Pro:kmeans alexanderbuccheri$ cmake --install serial-cmake-build-debug
-- Install configuration: "Debug"
-- Installing: /Users/alexanderbuccheri/Codes/kmeans/install/lib/libkmeans.1.0.0.dylib
```

AlexBuccheri replied:
Otherwise, what variable am I supposed to use?

LecrisUT replied:
It's not the variable it's the path that you should put it in  `/usr/...` have a defined meaning. There is no standard for `modules` so 🤷‍♂️. But use relative path installation.

Of course it should be an editable variable but I haven't got to the options part of the grill

AlexBuccheri committed:
Specify build and install include directories for lib_kmeans
Add Config.cmake.in template and install the Config.cmake file.

AlexBuccheri commented on cmake/kmeansConfig.cmake.in:
```diff
@@ -0,0 +1,11 @@
+find_dependency(OpenMP REQUIRED)
+find_dependency(MPI)
+# Include the Targets file to bring in the exported targets
+include(@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_LIBDIR@/cmake/kmeans/KmeansTargets.cmake)
```
This looks sketchy to me, but if I don't include it, I have to add 

`include(${kmeans_DIR}/kmeansTargets.cmake)
`
after find_package(), in the consumer, which is clearly a hack

LecrisUT replied:
`find_dependecy` are optional and only if the linkage is `PUBLIC/INTERFACE`, which it isn't.

Use `${CMAKE_CURRENT_LIST_DIR}` otherwise it's no longer relocatable

AlexBuccheri replied:
So, remove `find_dependency` cals?

Replace `@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_LIBDIR@` with `@{CMAKE_CURRENT_LIST_DIR}@`?

LecrisUT replied:
`${CMAKE_CURRENT_LIST_DIR} ` it **must not** be evaluated at configure time, but at the execution time (of `find_package`)

AlexBuccheri commented on CMakeLists.txt:
```diff
@@ -146,3 +170,76 @@ target_link_libraries(test_kmeans PRIVATE lib_kmeans fortuno_wrapper common_flag
+# cmake export files
+include(CMakePackageConfigHelpers)
+
+configure_package_config_file(cmake/kmeansConfig.cmake.in "${CMAKE_CURRENT_BINARY_DIR}/kmeansConfig.cmake"
+                               INSTALL_DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/kmeans")
```
Added missing `Config.cmake`, required for consuming

AlexBuccheri commented on CMakeLists.txt:
```diff
@@ -146,3 +170,76 @@ target_link_libraries(test_kmeans PRIVATE lib_kmeans fortuno_wrapper common_flag
+    DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/kmeans"
+)
+
+# Comment from Cristian on using .pc file:
```
Opting to leave .pc for now

AlexBuccheri committed:
Change module installation to relative path

AlexBuccheri committed:
Switch include of Targets to ${CMAKE_CURRENT_LIST_DIR}, else it's not portable
